### PR TITLE
fix: Add exact-match search on variable name in joint declarations

### LINKED ISSUE #522: BestFitScanner doesn't handle joint declarations well (affects DeadStoreProcessor (rule 1854))
Maybe you already have small reproducible examples and I am not sure if this is even related but running `java -jar sorald-0.2.0-jar-with-dependencies.jar repair --source=Test.java --rule-key=1854` (Unused assignments should be removed) on

```java
public class Test {
    public static void main(String[] args) {
        String x = "hello", y = "world";
        System.out.println(x);
    }
}
```
yields:
```java
public class Test {
    public static void main(String[] args) {String y = "world";
        String x;
        System.out.println(x);
    }
}
```

_Originally posted by @lyxell in https://github.com/SpoonLabs/sorald/issues/337#issuecomment-834429422_

### PR DESCRIPTION
**Pull request by slarse**
Fix #522 

This PR extends the best fits scanner with a post filter that runs on all position-matched candidates. Currently, this post filter is only used to improve matching between rule violations and jointly declared variables.

@lyxell I know we discussed hamming distance, but I figured an exact match should be equally viable here. It may be that we find corner cases where the entire identifier can't be extracted, but I couldn't come up with one. Feel free to suggest more test cases and comment on the implementation!

### CODE REVIEW

**Comment by slarse** [comment]
> Hopefully it should suffice to look for a comma as well?

Good catch! We can be super prudent and use [`Character.isJavaIdentifierPart`](https://docs.oracle.com/javase/8/docs/api/java/lang/Character.html#isJavaIdentifierPart-char-) (which is slightly more permissive than [`Character.isJavaIdentifierStart`](https://docs.oracle.com/javase/8/docs/api/java/lang/Character.html#isJavaIdentifierStart-char-) in allowing digits, but it should work fine).

File: src/main/java/sorald/sonar/BestFitScanner.java [hunk 1]
```diff
@@ -167,6 +171,45 @@ private boolean canRepair(E element) {
...
+     * violation.
+     */
+    private static String getPrecedingSymbol(RuleViolation violation, CtCompilationUnit cu) {
+        String cuSource = cu.getOriginalSourceCode();
+        int searchStartPos =
+                calculateSourcePos(
+                        violation.getStartLine(),
+                        violation.getStartCol(),
+                        cu.getLineSeparatorPositions());
+        int symbolEndPos = reverseSearch(cuSource, searchStartPos, Character::isJavaIdentifierPart);
```
Comment (lyxell) [inline src/main/java/sorald/sonar/BestFitScanner.java:199]: Maybe naming these `identifierStartPos` and `identifierEndPos` would be more intuitive?

**Review by lyxell — CHANGES_REQUESTED** [review]

File: src/main/java/sorald/sonar/BestFitScanner.java [hunk 2]
```diff
@@ -167,6 +171,45 @@ private boolean canRepair(E element) {
...
+        int symbolStartPos =
+                reverseSearch(cuSource, symbolEndPos, c -> !Character.isJavaIdentifierPart(c)) + 1;
+        return cuSource.substring(symbolStartPos, symbolEndPos + 1);
+    }
+
+    private static int reverseSearch(String s, int startIdx, Predicate<Character> predicate) {
```
[the remark this answers is not in the document]
Comment (lyxell) [reply src/main/java/sorald/sonar/BestFitScanner.java:205]: I was just referring to the name of the method, I think using the predicate is a good idea.

File: src/main/java/sorald/sonar/BestFitScanner.java (hunk 1 above)
Comment (lyxell) [reply src/main/java/sorald/sonar/BestFitScanner.java:199]: :+1:

File: src/main/java/sorald/sonar/BestFitScanner.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (slarse) [reply src/main/java/sorald/sonar/BestFitScanner.java:205]: Oh, darn, I entirely misread your comment, now I totally see what you mean. The negation of the predicate is super confusing in combination with the method name. I don't find that nitpicking, that's just bad design (probably a product of my refactoring a loop out of the original method and not thinking twice).

Just a note: I might have understood what you meant if you'd commented on the entire method s.t. the negated predicate was actually in view (you can select multiple lines to comment by clicking-and-dragging in the margin).

[the remark this answers is not in the document]
Comment (slarse) [reply src/main/java/sorald/sonar/BestFitScanner.java:205]: Or wait, am I just tired now? `while predicate is not true keep searching`? Does that not make sense?

[the remark this answers is not in the document]
Comment (lyxell) [reply src/main/java/sorald/sonar/BestFitScanner.java:205]: Arguably, from reading the signature I still don't think it is entirely clear whether the method searches *until* the predicate holds or *while* the predicate holds, even though I would probably bet on the former. But if you think it's clear then it's all good.

[the remark this answers is not in the document]
Comment (lyxell) [reply src/main/java/sorald/sonar/BestFitScanner.java:205]: I mean, I think it is already *kind of* clear, but I do think *reverseFind* would be even clearer.

**Comment by slarse** [comment]
Codecov upload is currently failing. They'll probably sort it out during the day so I'll just wait a bit and re-run workflows after lunch.

@lyxell Thanks for the reviews, feel free to have a final look, I'll ping someone to merge when CI passes again.

**Comment by lyxell** [comment]
Looks good to me!

**Comment by slarse** [comment]
@fermadeiral Ready for merge.

**Comment by fermadeiral** [comment]
Nice job! Thanks @slarse and @lyxell!
